### ISSUE TITLE
On-board array module, or at least proper is_array function

### ISSUE THREAD

**Issue by sharonovd** [tarantool/tarantool#2765]
I am currently using is_array from here: https://github.com/mpx/lua-cjson/blob/db12267686af80f0a3643897d09016c137715c8a/lua/cjson/util.lua

Looks like microlight (https://github.com/stevedonovan/Microlight) has an attention-worthy implementation of arrays generally.

**Comment by rtsisyk** [comment]
Hi, there is already an isarray() implementation on-board used internally by msgpack/json/yaml. Shall we export it as table.isarray()?

**Comment by Totktonada** [comment]
@rtsisyk How this internal isarray() implementation behaves (what is considered as an array)?

**Comment by sharonovd** [comment]
Strange - I don't see it in the sources (I honestly hope you're not talking about one-liners from json.test.lua etc.)

**Comment by rtsisyk** [comment]
luaL_serialize in utils.c

**Comment by Totktonada** [comment]
@rtsisyk I guess you mean [lua_field_inspect_table](https://github.com/tarantool/tarantool/blob/5a8243d538beedce85420096221abc7527860eb7/src/lua/utils.c#L377)(). The approach looks similar to lua-cjson one, but it’s possible to set type manually using __serialize metatable field (set it to 'seq' or 'map'). Is marking a table array or map documented somewhere?

**Comment by rtsisyk** [comment]
@Totktonada yeah, __serialize support is important. Let's export  lua_field_inspect_table() via table.isarray() (without underscore).

**Comment by rtsisyk** [comment]
Discussed.
There is no way to implement universal is_array() in Lua. I tried different approaches and all of them can fail under certain conditions. The final solution was to add  `encode_sparse_ratio` and `encode_sparse_safe` serializing options (msgpack/json/yaml) and copy-cat [sparse array](https://www.kyne.com.au/~mark/software/lua-cjson-manual.html#encode_sparse_array) detection logic from cjson. I don't think that we want this logic in built-in `table` module. However, I keep this feature request open until new `array` module is implemented or imported.

See also https://github.com/tarantool/tarantool/issues/1167#issuecomment-174489558

**Comment by knazarov** [comment]
Please like my post if you think that is_array should just check if table is a contiguous sequence of elements with numeric keys from 1 to n. (without holes)

In my opinion it is the most practical and useful definition.

**Comment by mtrempoltsev** [comment]
I think array cannot be zero size (see C lang as example)

**Comment by Totktonada** [comment]
NB: We can add it to https://github.com/tarantool/checks

**Comment by rosik** [comment]
I dislike, because your definition of array differs from other languages and therefore may be misleading.

In C++ I can store `nullptr` in `std::vector<void*>`

In Python `len([1, None, 3, None, None]) == 5`

**Comment by Totktonada** [comment]
See also https://github.com/openresty/luajit2/commit/5e4cf2be2652b3639c79d50a0412b9ae7de6c68a and especially the discussion in https://github.com/openresty/luajit2/commit/5e4cf2be2652b3639c79d50a0412b9ae7de6c68a#r32252602

**Comment by Totktonada** [comment]
See also https://wiki.php.net/rfc/is_list. However take into account that a hash map in PHP can store `null` values (like in C++, Python, but in opposite to Lua).

```php
php > $h = [1, null, 3];
php > var_dump($h);
array(3) {
  [0]=>
  int(1)
  [1]=>
  NULL
  [2]=>
  int(3)
}
php > var_dump(array_is_list($h));
bool(true)
```
